### PR TITLE
[Clang][Sema] Fix crash when 'this' is used in a dependent class scope function template specialization that instantiates to a static member function

### PR DESCRIPTION
**Pull request by sdkrystian**
This patch fixes a crash that happens when '`this`' is referenced (implicitly or explicitly) in a dependent class scope function template specialization that instantiates to a static member function. For example ([godbolt](https://godbolt.org/z/4c4d4M1nf)):
```cpp
template<typename T>
struct A 
{
    template<typename U>
    static void f();

    template<>
    void f<int>() 
    {
        this; // causes crash during instantiation
    }
};

template struct A<int>;
```
This happens because during instantiation of the function body, `Sema::getCurrentThisType` will return a null `QualType` which we rebuild the `CXXThisExpr` with. A similar problem exists for implicit class member access expressions in such contexts (which shouldn't really happen within templates anyways per [[class.mfct.non.static] p2](http://eel.is/c++draft/class.mfct.non.static#2), but changing that is non-trivial). This patch fixes the crash by building `UnresolvedLookupExpr`s instead of `MemberExpr`s for these implicit member accesses, which will then be correctly rebuilt as `MemberExpr`s during instantiation.

### CODE REVIEW

**Comment by sdkrystian** [comment]
Still need to add a release note

File: clang/lib/Sema/SemaExprMember.cpp [hunk 1]
```diff
@@ -91,10 +95,20 @@ static IMAKind ClassifyImplicitMemberAccess(Sema &SemaRef,
...
+    if (!MD->isStatic() && !MD->isExplicitObjectMemberFunction())
+      isStaticOrExplicitContext = false;
+
+    // A dependent class scope function template explicit specialization
+    // which has no explicit object parameter and is declared without 'static'
+    // could instantiate to either a static or non-static member function.
+    if (MD->getDependentSpecializationInfo() && !MD->isStatic() &&
+        !MD->isExplicitObjectMemberFunction())
+      couldInstantiateToStatic = true;
+  }
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaExprMember.cpp:111]:
```suggestion
  bool couldInstantiateToStatic = false;
  bool isStaticOrExplicitContext = SemaRef.CXXThisTypeOverride.isNull();

  if (auto *MD = dyn_cast<CXXMethodDecl>(DC)) {
    if (MD->isImplicitObjectMemberFunction())
      isStaticOrExplicitContext = false;

    // A dependent class scope function template explicit specialization
    // which has no explicit object parameter and is declared without 'static'
    // could instantiate to either a static or non-static member function.
    if (MD->getDependentSpecializationInfo() && !isStaticOrExplicitContext)
      couldInstantiateToStatic = true;
  }
```

**Review by cor3ntin — COMMENTED** [review]
This looks correct, I think. I would prefer if the NFC changes were done as a (set of) separate commit though

File: clang/lib/Sema/SemaExprMember.cpp (hunk 1 above)
Comment (sdkrystian) [reply clang/lib/Sema/SemaExprMember.cpp:111]: @cor3ntin I simplified a bit further:
```cpp
bool couldInstantiateToStatic = false;
bool isStaticOrExplicitContext = SemaRef.CXXThisTypeOverride.isNull();

if (auto *MD = dyn_cast<CXXMethodDecl>(DC)) {
  if (MD->isImplicitObjectMemberFunction()) {
    isStaticOrExplicitContext = false;
    // A dependent class scope function template explicit specialization
    // that is neither declared 'static' nor with an explicit object
    // parameter could instantiate to a static or non-static member function.
    couldInstantiateToStatic = MD->getDependentSpecializationInfo();
  }
}
```

**Comment by sdkrystian** [comment]
@cor3ntin Could you elaborate as to which NFC changes should be split into separate commits? There aren't any changes included that are needed for the patch..

**Comment by sdkrystian** [comment]
Ping @erichkeane :)

**Comment by erichkeane** [comment]
I'm OK with this, but would like @cor3ntin to take another look since he engaged earlier.

**Review by cor3ntin — APPROVED** [review]
LGTM

**Comment by rupprecht** [comment]
This commit appears to regress an example like this:

```c++
template <typename T>
struct Foo {
  template <typename X>
  int bar(X x) {
    return 0;
  }

  template <>
  int bar(int x) {
    return bar(5.0);
  }
};

void call() {
  Foo<double> f;
  f.bar(1);
}
```

Used to compile, now results in an error:
```
PR87541.cpp:10:12: error: call to non-static member function without an object argument
   10 |     return bar(5.0);
      |            ^~~
PR87541.cpp:16:5: note: in instantiation of function template specialization 'Foo<double>::bar<int>' requested here
   16 |   f.bar(1);
      |     ^
```

Is that intended?

**Comment by sdkrystian** [comment]
@rupprecht That's unintended, will revert and address it.

**Pull request merged**